Bitcode incompatibility for XCTest dependent frameworks on tvOS and watchOS

I'd like to discuss the possibility of better Carthage support for test frameworks on platforms that require bitcode.

* `carthage version`: 0.16
* `xcodebuild -version`: Xcode 7.3, Build version 7D175

## Situation

Apps built for a tvOS device are required to be built with bitcode. This causes problems for frameworks like Quick and Nimble that link to XCTest as it does not contain bitcode.

**Carthage Output**
```
$ carthage build --no-skip-current --platform tvos
*** xcodebuild output can be found in /var/folders/96/rcjvzrkn5lzfl0_krn8c151h0000gn/T/carthage-xcodebuild.TE6TlH.log
*** Building scheme "Quick-tvOS" in Quick.xcworkspace
** BUILD FAILED **


The following build commands failed:
	Ld /Volumes/DerivedData/Quick-dshynvhuyaarriayewbsnvjxqrux/Build/Products/Release-appletvos/Quick.framework/Quick normal arm64
(1 failure)
ld: warning: Auto-Linking supplied '/Applications/Xcode7.3.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/Library/Frameworks/XCTest.framework/XCTest', '/Applications/Xcode7.3.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/Library/Frameworks/XCTest.framework/XCTest' does not contain bitcode. You must rebuild it with bitcode enabled (Xcode setting ENABLE_BITCODE) or obtain an updated library from the vendor
clang: error: linker command failed with exit code 1 (use -v to see invocation)
A shell task (/usr/bin/xcrun xcodebuild -workspace /Users/phatblat/dev/ios/pods/Quick/Quick.xcworkspace -scheme Quick-tvOS -configuration Release -sdk appletvos ONLY_ACTIVE_ARCH=NO BITCODE_GENERATION_MODE=bitcode CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY= CARTHAGE=YES clean build) failed with exit code 65:
** BUILD FAILED **


The following build commands failed:
	Ld /Volumes/DerivedData/Quick-dshynvhuyaarriayewbsnvjxqrux/Build/Products/Release-appletvos/Quick.framework/Quick normal arm64
(1 failure)
```

Whenever we get [XCTest on watchOS](https://openradar.appspot.com/21760513) this will be an issue there as well.

These frameworks are not intended to be included with apps submitted to the app store. It would be nice if they could be included in device builds for local development testing. However, for platforms other than OS X, their primary environment is the various platform simulators.

## Affected Frameworks

- Quick/Nimble#213
- Quick/Quick#422

This is a total Catch-22 situation that Apple has put us in and I'm hoping that Carthage can make it a little less painful.

There are [workarounds](https://github.com/Quick/Nimble/issues/213#issuecomment-202496390) to this issue. Also, if Apple ever relaxes the bitcode requirement on tvOS and watchOS or adds bitcode to the XCTest framework, this will become a non-issue.

## Proposal

It is incredibly convenient for development that Carthage builds fat binaries with both device and simulator slices. However, when building a unit testing framework for a platform that requires bitcode this is a bit of a problem.

Since there is not presently a way to tell Carthage to only build for simulator architectures, what if the framework project's `ENABLE_BITCODE` setting of `NO` plus the combination of a platform that requires bitcode (`tvOS` and `watchOS`) were to disable generating the device architecture slices?

### Example tvOS Framework

- platform: `tvOS`
- `ENABLE_BITCODE`: `YES`
- generated slices: `x84_64`, `arm64`

### Example tvOS Test Framework

- platform: `tvOS`
- `ENABLE_BITCODE`: `NO`
- generated slices: `x84_64`

### Example watchOS Framework

- platform: `watchOS`
- `ENABLE_BITCODE`: `YES`
- generated slices: `i386`, `armv7k`

### Example watchOS Test Framework

- platform: `watchOS`
- `ENABLE_BITCODE`: `NO`
- generated slices: `i386`

Setting `ENABLE_BITCODE` to `NO` in most cases is an error on these platforms so I could see Carthage generating some sort of "are you sure?" warning. This approach seems simpler to me than something fancy like detecting that the framework links to XCTest or some other framework that isn't built with bitcode.

--- a/Source/CarthageKit/Xcode.swift
+++ b/Source/CarthageKit/Xcode.swift
@@ -969,8 +969,10 @@ public func buildScheme(scheme: String, withConfiguration configuration: String,
 			let platform = sdk.platform
 
 			if var sdks = sdksByPlatform[platform] {
-				sdks.append(sdk)
-				sdksByPlatform.updateValue(sdks, forKey: platform)
+				if sdks.indexOf(sdk) == nil {
+					sdks.append(sdk)
+					sdksByPlatform.updateValue(sdks, forKey: platform)
+				}
 			} else {
 				sdksByPlatform[platform] = [ sdk ]
 			}

--- a/Source/CarthageKit/Xcode.swift
+++ b/Source/CarthageKit/Xcode.swift
@@ -964,17 +964,15 @@ public func buildScheme(scheme: String, withConfiguration configuration: String,
 				}
 				.map { _ in sdk }
 		}
-		.reduce([:]) { (sdksByPlatform: [Platform: [SDK]], sdk: SDK) in
+		.reduce([:]) { (sdksByPlatform: [Platform: Set<SDK>], sdk: SDK) in
 			var sdksByPlatform = sdksByPlatform
 			let platform = sdk.platform
 
 			if var sdks = sdksByPlatform[platform] {
-				if sdks.indexOf(sdk) == nil {
-					sdks.append(sdk)
-					sdksByPlatform.updateValue(sdks, forKey: platform)
-				}
+				sdks.insert(sdk)
+				sdksByPlatform.updateValue(sdks, forKey: platform)
 			} else {
-				sdksByPlatform[platform] = [ sdk ]
+				sdksByPlatform[platform] = Set(arrayLiteral: sdk)
 			}
 
 			return sdksByPlatform
@@ -984,7 +982,7 @@ public func buildScheme(scheme: String, withConfiguration configuration: String,
 				fatalError("No SDKs found for scheme \(scheme)")
 			}
 
-			let values = sdksByPlatform.map { ($0, $1) }
+			let values = sdksByPlatform.map { ($0, Array($1)) }
 			return SignalProducer(values: values)
 		}
 		.flatMap(.Concat) { platform, sdks -> SignalProducer<(Platform, [SDK]), CarthageError> in

--- a/Source/CarthageKit/Xcode.swift
+++ b/Source/CarthageKit/Xcode.swift
@@ -950,24 +950,19 @@ public func buildScheme(scheme: String, withConfiguration configuration: String,
 	}
 
 	return BuildSettings.SDKsForScheme(scheme, inProject: project)
-		.flatMap(.Concat) { sdk -> SignalProducer<(SDK, BuildSettings), CarthageError> in
+		.flatMap(.Concat) { sdk -> SignalProducer<SDK, CarthageError> in
 			var argsForLoading = buildArgs
 			argsForLoading.sdk = sdk
 
-			return SignalProducer(value: sdk)
-				.zipWith(BuildSettings.loadWithArguments(argsForLoading))
-		}
-		.filter { sdk, settings in
-			// Filter out SDKs that require bitcode when bitcode is disabled in
-			// project settings. This is necessary for testing frameworks, which
-			// must add a User-Defined setting of ENABLE_BITCODE=NO.
-			if settings.bitcodeEnabled.value != true && [.tvOS, .watchOS].contains(sdk) {
-				return false
-			}
-			return true
-		}
-		.flatMap(.Concat) { sdk, _ -> SignalProducer<SDK, CarthageError> in
-			return SignalProducer(value: sdk)
+			return BuildSettings
+				.loadWithArguments(argsForLoading)
+                .filter { settings in
+                    // Filter out SDKs that require bitcode when bitcode is disabled in
+                    // project settings. This is necessary for testing frameworks, which
+                    // must add a User-Defined setting of ENABLE_BITCODE=NO.
+					return settings.bitcodeEnabled.value == true || ![.tvOS, .watchOS].contains(sdk)
+				}
+				.map { _ in sdk }
 		}
 		.reduce([:]) { (sdksByPlatform: [Platform: [SDK]], sdk: SDK) in
 			var sdksByPlatform = sdksByPlatform

--- a/Source/CarthageKit/Xcode.swift
+++ b/Source/CarthageKit/Xcode.swift
@@ -36,7 +36,7 @@ public enum ProjectLocator: Comparable {
 			return URL
 		}
 	}
-	
+
 	/// The number of levels deep the current object is in the directory hierarchy.
 	public var level: Int {
 		return fileURL.pathComponents!.count - 1
@@ -63,7 +63,7 @@ public func <(lhs: ProjectLocator, rhs: ProjectLocator) -> Bool {
 	guard leftLevel == rightLevel else {
 		return leftLevel < rightLevel
 	}
-	
+
 	// Prefer workspaces over projects.
 	switch (lhs, rhs) {
 	case (.Workspace, .ProjectFile):
@@ -523,7 +523,7 @@ public struct BuildSettings {
 							currentTarget = (line as NSString).substringWithRange(targetRange)
 							return
 						}
-						
+
 						let trimSet = NSCharacterSet.whitespaceAndNewlineCharacterSet()
 						let components = line.characters
 							.split(1) { $0 == "=" }
@@ -950,6 +950,25 @@ public func buildScheme(scheme: String, withConfiguration configuration: String,
 	}
 
 	return BuildSettings.SDKsForScheme(scheme, inProject: project)
+		.flatMap(.Concat) { sdk -> SignalProducer<(SDK, BuildSettings), CarthageError> in
+			var argsForLoading = buildArgs
+			argsForLoading.sdk = sdk
+
+			return SignalProducer(value: sdk)
+				.zipWith(BuildSettings.loadWithArguments(argsForLoading))
+		}
+		.filter { sdk, settings in
+			// Filter out SDKs that require bitcode when bitcode is disabled in
+			// project settings. This is necessary for testing frameworks, which
+			// must add a User-Defined setting of ENABLE_BITCODE=NO.
+			if settings.bitcodeEnabled.value != true && [.tvOS, .watchOS].contains(sdk) {
+				return false
+			}
+			return true
+		}
+		.flatMap(.Concat) { sdk, _ -> SignalProducer<SDK, CarthageError> in
+			return SignalProducer(value: sdk)
+		}
 		.reduce([:]) { (sdksByPlatform: [Platform: [SDK]], sdk: SDK) in
 			var sdksByPlatform = sdksByPlatform
 			let platform = sdk.platform
@@ -1274,7 +1293,7 @@ private func stripBinary(binaryURL: NSURL, keepingArchitectures: [String]) -> Si
 public func copyProduct(from: NSURL, _ to: NSURL) -> SignalProducer<NSURL, CarthageError> {
 	return SignalProducer<NSURL, CarthageError>.attempt {
 		let manager = NSFileManager.defaultManager()
-		
+
 		// This signal deletes `to` before it copies `from` over it.
 		// If `from` and `to` point to the same resource, there's no need to perform a copy at all
 		// and deleting `to` will also result in deleting the original resource without copying it.
@@ -1286,7 +1305,7 @@ public func copyProduct(from: NSURL, _ to: NSURL) -> SignalProducer<NSURL, Carth
 			from.absoluteURL == to.absoluteURL {
 				return .Success(to)
 		}
-		
+
 		do {
 			try manager.createDirectoryAtURL(to.URLByDeletingLastPathComponent!, withIntermediateDirectories: true, attributes: nil)
 		} catch let error as NSError {

--- a/Source/CarthageKit/Xcode.swift
+++ b/Source/CarthageKit/Xcode.swift
@@ -956,10 +956,10 @@ public func buildScheme(scheme: String, withConfiguration configuration: String,
 
 			return BuildSettings
 				.loadWithArguments(argsForLoading)
-                .filter { settings in
-                    // Filter out SDKs that require bitcode when bitcode is disabled in
-                    // project settings. This is necessary for testing frameworks, which
-                    // must add a User-Defined setting of ENABLE_BITCODE=NO.
+				.filter { settings in
+					// Filter out SDKs that require bitcode when bitcode is disabled in
+					// project settings. This is necessary for testing frameworks, which
+					// must add a User-Defined setting of ENABLE_BITCODE=NO.
 					return settings.bitcodeEnabled.value == true || ![.tvOS, .watchOS].contains(sdk)
 				}
 				.map { _ in sdk }

